src/derived.rs

```diff
@@ -501,6 +515,37 @@ where
     }
 }
 
+struct PanicGuard<'db, DB, Q>
+where
+    DB: Database + 'db,
+    Q: QueryFunction<DB>,
+{
+    map: &'db RwLock<FxHashMap<Q::Key, QueryState<DB, Q>>>,
+    key: &'db Q::Key,
+}
+
+impl<'db, DB, Q> PanicGuard<'db, DB, Q>
+where
+    DB: Database + 'db,
+    Q: QueryFunction<DB>,
+{
+    fn new(map: &'db RwLock<FxHashMap<Q::Key, QueryState<DB, Q>>>, key: &'db Q::Key) -> Self {
+        Self { map, key }
+    }
+}
+
+impl<'db, DB, Q> Drop for PanicGuard<'db, DB, Q>
+where
+    DB: Database + 'db,
+    Q: QueryFunction<DB>,
+{
+    fn drop(&mut self) {
+        let map = self.map.upgradable_read();
+        let mut map = RwLockUpgradableReadGuard::upgrade(map);
+        let _ = map.remove(self.key);
```

Comment:
To handle the parallel case, we would basically just *inspect* this returned value, much as we do in `overwrite_placeholder`:

https://github.com/salsa-rs/salsa/blob/bfc56d25919c5df7271f2c6421c082d6b93b003e/src/derived.rs#L459

This consists of two steps:

First, with the lock held, we check if there are any waiters and invoke `unblock_queries_blocked_on_self` to update salsa's internal data structures:

https://github.com/salsa-rs/salsa/blob/bfc56d25919c5df7271f2c6421c082d6b93b003e/src/derived.rs#L474-L481

Next, once lock is released, we sent messages:

```rust
https://github.com/salsa-rs/salsa/blob/bfc56d25919c5df7271f2c6421c082d6b93b003e/src/derived.rs#L490-L492
```

Before you tackle this, though, it's a good idea to add some test cases that (ought to...) deadlock, to make sure we are testing what we think we are.